### PR TITLE
#438

### PR DESCRIPTION
**Pull request by HDouss**
Resolving #438 .
Added a puzzle to implement chunked bodies.

### CODE REVIEW

**Comment by davvd** [comment]
@hdouss Thanks for your pull request, let me find someone who can review it

**Comment by davvd** [comment]
@dmzaytsev it's yours, please review

File: src/main/java/org/takes/rq/RqMultipart.java [hunk 1]
```diff
@@ -418,19 +418,21 @@ public Fake(final Request req, final Request... dispositions)
...
                     @Override
                     public Iterable<String> head() throws IOException {
-                        return new RqWithHeader(
+                        return new RqWithHeaders(
                             req,
-                            // @checkstyle MultipleStringLiteralsCheck (1 line)
-                            "Content-Type",
                             String.format(
-                                "multipart/form-data; boundary=%s",
+                                "Content-Type:multipart/form-data; boundary=%s",
```
Comment (dmzaytsev) [inline src/main/java/org/takes/rq/RqMultipart.java]: @HDouss I think we need one space before the header value

File: src/main/java/org/takes/rq/RqMultipart.java [hunk 2]
```diff
@@ -418,19 +418,21 @@ public Fake(final Request req, final Request... dispositions)
...
                                 RqMultipart.Fake.BOUNDARY
+                            ),
+                            String.format(
+                                "Content-Length:%s",
```
Comment (dmzaytsev) [inline src/main/java/org/takes/rq/RqMultipart.java]: @HDouss and here

File: src/test/java/org/takes/rq/RqChunkTest.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+            new RqFake(
+                Arrays.asList(
+                    "GET /h?a=1",
+                    "Host: www.example.com",
+                    CHUNKED_HEADER
+                ),
+                Joiner.on(CRLF).join(
+                    length,
+                    data,
+                    END_OF_CHUNK
```
Comment (dmzaytsev) [inline src/test/java/org/takes/rq/RqChunkTest.java]: @HDouss I think you need one CRLF after last chunk
https://www.w3.org/Protocols/rfc2616/rfc2616-sec3.html
#### 3.6.1 Chunked Transfer Coding
```
       Chunked-Body   = *chunk
                        last-chunk
                        trailer
                        CRLF
....
       last-chunk     = 1*("0") [ chunk-extension ] CRLF
```

File: src/test/java/org/takes/rq/RqChunkTest.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+    public void readsManyChunks() throws IOException {
+        final String first = "Takes is";
+        final String second = "a true object-";
+        final String third = "oriented framework";
+        final String data = first + second + third;
+        final Integer length = data.length();
+        final InputStream stream = new RqChunk(
+            new RqFake(
+                Arrays.asList(
+                    "GET /h?a=2",
```
Comment (dmzaytsev) [inline src/test/java/org/takes/rq/RqChunkTest.java]: @HDouss we have invalid chunk here 
see above

File: src/test/java/org/takes/rq/RqChunkTest.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+            new RqFake(
+                Arrays.asList(
+                    "GET /h?a=3",
+                    "Host: c.example.com",
+                    CHUNKED_HEADER
+                ),
+                Joiner.on(CRLF).join(
+                    length + ignored,
+                    data,
+                    END_OF_CHUNK
```
Comment (dmzaytsev) [inline src/test/java/org/takes/rq/RqChunkTest.java]: @HDouss and here missed trailing CRLF

**Comment by dmzaytsev** [comment]
@HDouss  please see my comments

File: src/test/java/org/takes/rq/RqChunkTest.java (hunk 4 above)
Comment (HDouss) [inline src/test/java/org/takes/rq/RqChunkTest.java]: @dmzaytsev You mean the trailing "CRLF" ?

**Comment by HDouss** [comment]
@dmzaytsev Thanks. Done !

File: src/test/java/org/takes/rq/RqChunkTest.java (hunk 4 above)
Comment (dmzaytsev) [inline src/test/java/org/takes/rq/RqChunkTest.java]: @HDouss exactly

**Comment by dmzaytsev** [comment]
@HDouss thank you

**Comment by dmzaytsev** [comment]
@rultor merge

**Comment by HDouss** [comment]
@yegor256 Please confirm the merge.

**Comment by yegor256** [comment]
@rultor merge

**Comment by davvd** [comment]
@elenavolokhova please, help us to evaluate quality here, as per [par.24](http://at.teamed.io/policy.html#24)

**Comment by davvd** [comment]
@rultor deploy

**Comment by elenavolokhova** [comment]
@davvd Looks good!

**Comment by davvd** [comment]
> @davvd Looks good!

@elenavolokhova thanks, got it, great to see high-quality tickets here

**Comment by davvd** [comment]
@dmzaytsev paied 10 mins to @elenavolokhova for QA review (payment ID is `75590483`)

Thanks so much! Your account was topped up for **24 mins** (transaction ID is `75590777`, the task took 87 hours and 32 mins)

extra minutes for review comments (c=9)

+24 added to your rating, current score is: [+2634](http://www.netbout.com/b/35166?open=rating)

**Pull request merged**